src/clp_logging/handlers.py

```diff
@@ -792,3 +800,168 @@ def __init__(
+    # override
+    def setStream(self, stream: IO[bytes]) -> Optional[IO[bytes]]:
+        if stream is self._ostream:
+            return None
+        old_stream: IO[bytes] = self._ostream
+        self._ostream = stream
+        # TODO: The following call will close the old stream. However, `logging.StreamHandler`'s
+        # implementation will only flush the stream but leave it opened. To support this behaviour,
+        # we need `clp_ffi_py.ir.Serializer` to allow closing the serializer without closing the
+        # underlying output stream.
+        self._init_new_serializer(stream)
+        return old_stream
+
+    # override
+    def close(self) -> None:
+        if self._serializer is None:
+            return
+        if self._loglevel_timeout:
+            self._loglevel_timeout.timeout()
+        # NOTE: Closing the serializer will ensure that any buffered results are flushed and the
+        # underlying output stream is properly closed.
+        self._serializer.close()
+        self._serializer = None
+        super().close()
+
+    def _is_closed(self) -> bool:
+        return self._serializer is None
+
+    def _init_new_serializer(self, stream: IO[bytes]) -> None:
+        """
+        Initializes a new serializer that writes to the given stream.
+
+        :param stream: The stream that the underlying serializer will write to.
+        """
+        cctx: ZstdCompressor = ZstdCompressor()
+        self._serializer = Serializer(
+            cctx.stream_writer(stream) if self._enable_compression else stream
+        )
+
+    def _write(self, record: logging.LogRecord) -> None:
+        """
+        Writes the log event into the underlying serializer.
+
+        :param record: The log event to serialize.
+        :raise IOError: If the handler has been already closed.
+        :raise TypeError: If `record.msg` is not a Python dictionary.
+        """
+        if self._is_closed():
+            raise IOError("The handler has been closed.")
+
+        if not isinstance(record.msg, dict):
+            raise TypeError("The log msg must be a valid Python dictionary.")
+
+        curr_ts: int = floor(time.time() * 1000)
+
+        if self._loglevel_timeout is not None:
+            self._loglevel_timeout.update(record.levelno, curr_ts, self._log_level_timeout_callback)
+
+        self._serialize_kv_pair_log_event(
+            self._auto_generated_kv_pairs_buf.generate(curr_ts, self._tz, record), record.msg
+        )
+
+    def _serialize_kv_pair_log_event(
+        self, auto_generated_kv_pairs: Dict[str, Any], user_generated_kv_pairs: Dict[str, Any]
+    ) -> None:
+        """
+        :param auto_generated_kv_pairs: A dict of auto generated kv pairs.
+        :param user_generated_kv_pairs: A dict of user generated kv pairs.
+        """
+        log_event: Dict[str, Any] = {
+            AUTO_GENERATED_KV_PAIRS_KEY: auto_generated_kv_pairs,
+            USER_GENERATED_KV_PAIRS_KEY: user_generated_kv_pairs,
+        }
```

Comment:
Hm, I would've thought that before we have support for auto-generated kv-pairs in clp-ffi-py, we would make everything a user-generated kv-pair, except auto-generated kv-pairs would be prefixed with `@`. The current approach makes it so that the user has no way of knowing how to query their logs unless they know what we're doing here, right?